src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+            start_pattern: Regex::new("^[^\\s]").unwrap(),
+            condition_pattern: Regex::new("^[\\s]+from").unwrap(),
+            mode: Mode::ContinueThrough,
+            timeout: Duration::from_millis(10),
+        };
+        let expected = vec![concat!(
+            "foobar.rb:6:in `/': divided by 0 (ZeroDivisionError)\n",
+            "\tfrom foobar.rb:6:in `bar'\n",
+            "\tfrom foobar.rb:2:in `foo'\n",
+            "\tfrom foobar.rb:9:in `<main>'",
+        )];
+        run_and_assert(&lines, config, &expected);
+    }
+
+    #[test]
+    fn legacy() {
+        let lines = vec![
+            "INFO some usual line",
+            "INFO some other usual line",
+            "INFO first part",
+            "second part",
+            "last part",
+            "ERROR another normal message",
+            "ERROR finishing message",
+            "last part of the incomplete finishing message",
+        ];
+        let expected = vec![
+            "INFO some usual line",
+            "INFO some other usual line",
+            concat!("INFO first part\n", "second part\n", "last part"),
+            "ERROR another normal message",
+            concat!(
+                "ERROR finishing message\n",
+                "last part of the incomplete finishing message"
+            ),
+        ];
+
+        let stream = stream_from_lines(&lines);
+        let line_agg = LineAgg::new(
+            stream,
+            Config::for_legacy(
+                Regex::new("^(INFO|ERROR)").unwrap(), // example from the docs
+                10,
+            ),
+        );
+        let results = collect_results(line_agg);
+        assert_results(results, &expected);
+    }
+
+    // Test helpers.
+
+    fn stream_from_lines<'a>(
+        lines: &'a [&'static str],
+    ) -> impl Stream<Item = (Bytes, Filename), Error = ()> + 'a {
+        futures::stream::iter_ok::<_, ()>(
+            lines
+                .iter()
+                .map(|line| (Bytes::from_static(line.as_bytes()), "test.log".to_owned())),
+        )
+    }
+
+    fn collect_results<T>(line_agg: LineAgg<T>) -> Vec<(Bytes, Filename)>
+    where
+        T: Stream<Item = (Bytes, Filename), Error = ()>,
+    {
+        futures::future::Future::wait(futures::stream::Stream::collect(line_agg))
+            .expect("Failed to collect test results")
+    }
+
+    fn assert_results(actual: Vec<(Bytes, Filename)>, expected: &[&'static str]) {
+        let expected_mapped: Vec<(Bytes, Filename)> = expected
+            .iter()
+            .map(|line| (Bytes::from_static(line.as_bytes()), "test.log".to_owned()))
+            .collect();
+
+        assert_eq!(actual, expected_mapped);
+    }
+
+    fn run_and_assert(lines: &[&'static str], config: Config, expected: &[&'static str]) {
+        let stream = stream_from_lines(lines);
+        let line_agg = LineAgg::new(stream, config);
+        let results = collect_results(line_agg);
+        assert_results(results, expected);
+    }
```

Comment:
This looks like a good job of removing a pile of common boilerplate from the individual tests.